### PR TITLE
feat: add trainer padded batching

### PR DESCRIPTION
**Pull request by ShubyM**
Adds server side padded batching for forward_backward passes. Before we had code that looked like this: 

```python
for datum in data:
    loss = compute_loss(datum)
    loss.backward()

optimizer.step()
```
This means every datum does it's own small forward backward. To increase GPU utilization we can put multiple datums together in padded chunks and submit them all at once. To avoid OOMing, We can control how big an individual chunk can be with a new environment variable called ```OPEN_RL_TRAIN_TOKEN_BUDGET```. By default it's set 0 so we preserve the old behavior and make it opt-in.

### CODE REVIEW

File: src/server/trainer.py [hunk 1]
```diff
@@ -252,141 +252,170 @@ def forward_backward(self, data: list[Datum], loss_fn: str, loss_config: dict |
       self.peft_model.set_adapter(model_id)
 
     total_loss = 0.0
-    loss_fn_outputs = []
+    loss_fn_outputs: list[dict[str, Any] | None] = [None] * len(data)
 
-    # Ensure model is in train mode
     self.peft_model.train()
```
Comment (droot) [inline src/server/trainer.py:257]: that comment was helpful :)

File: src/server/trainer.py [hunk 2]
```diff
@@ -252,141 +252,170 @@ def forward_backward(self, data: list[Datum], loss_fn: str, loss_config: dict |
...
 
-    for datum in data:
-      # 1. Common Setup: Extract tokens and get logprobs
-      target_logprobs, targets_tensor, weights_tensor = self._get_logprobs(datum)
+    for batch in self._make_training_batches(data):
```
Comment (droot) [inline src/server/trainer.py:259]: it will be nice if we can maintain the same batch length the client expects us to. I haven't looked into it, but would be good to look up how batch sizes impact the stability of training runs and what are the things we should watch for.

File: src/server/trainer.py [hunk 3]
```diff
@@ -252,141 +252,170 @@ def forward_backward(self, data: list[Datum], loss_fn: str, loss_config: dict |
...
-    # Optional KL penalty against the reference policy.
-    # Uses the unbiased estimator (ratio - 1) - log(ratio), which is
-    # non-negative and zero when the policy matches the reference.
     kl_coeff = loss_config.get("kl_coeff", 0.0) if loss_config else 0.0
     if kl_coeff > 0:
       kl = (ratio - 1) - diff
       elementwise_objective = elementwise_objective - kl_coeff * kl
+    return -(elementwise_objective * weights)
+
+  def _reference_logprobs_and_advantages(self, aux_inputs: dict[str, torch.Tensor], loss_fn: str) -> tuple[torch.Tensor, torch.Tensor]:
```
Comment (droot) [inline src/server/trainer.py:415]: The change are very intense so not sure how well to review it. I wonder if we can just run some basis recipe tests to produce the loss runs etc to see runs are stable.

**Comment by droot** [comment]
@ShubyM overall this a good find. I somehow thought we were already doing it :)

File: src/server/trainer.py (hunk 3 above)
Comment (ShubyM) [reply src/server/trainer.py:415]: I can benchmark a few to ensure we get similar metrics. Will follow up

Comment (ShubyM) [reply src/server/trainer.py:415]: Confirmed on our recipes that our loss metrics match what we expect

**Review by droot — APPROVED** [review]
